applications: asset_tracker_v2: Add documentation for the cloud module (#6315)
nrfconnect/sdk-nrf: applications/asset_tracker_v2/doc/cloud_module.rst

This enables the cloud to issue FOTA updates while the device is in field to update the application and modem firmware.
For additional documentation on the various FOTA implementations, refer to the respective client library documentation linked to in :ref:`Integration layers <integration_layers>`.

Connection Awareness
====================

The cloud module implements connection awareness by maintaing an internal state that is based on
events from the modem module and callbacks from the :ref:`Cloud wrapper API <api_cloud_wrapper>`.

If at any point the module is disconnected, it will try to reconnect given that an LTE connection is still valid.
The number of reconnection attempts can be adjusted by configuring the :ref:`CONFIG_CLOUD_CONNECT_RETRIES <CONFIG_CLOUD_CONNECT_RETRIES>` option.
Reconnection is implemented with a binary backoff based on the following lookup table:

.. code-block:: c

   static struct cloud_backoff_delay_lookup backoff_delay[] = {
      { 32 }, { 64 }, { 128 }, { 256 }, { 512 },
      { 2048 }, { 4096 }, { 8192 }, { 16384 }, { 32768 },
      { 65536 }, { 131072 }, { 262144 }, { 524288 }, { 1048576 }

>    };

Reviewer: **suggestion:** use [source include](https://docutils.sourceforge.io/docs/ref/rst/directives.html#toc-entry-39)

Author: I didn't know that was possible? @peknis

Author: I'll keep it as is. It will leave some artifacts in the actual source code that I'm not completely sold on. Let's postpone this discussion for now.

CONFIG_CLOUD_CLIENT_ID - Configuration for providing a custom cloud client ID
   This option sets the custom client ID for the respective cloud service.
   For setting a custom client ID, you need to set :kconfig:`CONFIG_CLOUD_CLIENT_ID_USE_CUSTOM` to ``y``.

.. _CONFIG_CLOUD_CONNECT_RETRIES:

CONFIG_CLOUD_CONNECT_RETRIES - Configuration that sets the number of cloud reconnection attempts
   This option sets the number of times that a connection will be re-attempted upon a disconnect from the cloud service.

.. _mandatory_config:

Mandatory configurations
========================

To be able to use a supported cloud client library, a set of mandatory Kconfig options needs to be set.
These typically include the cloud service hostname and the security tag associated with the certificates used to establish a connection.
The aforementioned certificates needs to provisioned to the modem using the same security tag, prior to running the application.
For more information on how to setup a connection and provision certificates to the modem, see the documentation for the respective client library listed in :ref:`Integration layers <integration_layers>`.

.. note::

>    The :ref:`lib_nrf_cloud` library does not require any mandatory configurations to be set.

Reviewer 2: ```suggestion
   There are no mandatory configuration settings for the :ref:`lib_nrf_cloud` library.
```

>    The nRF9160 on the DK and Thingy91 comes preprovisioned with certificates required to establish a connection to nRF Cloud.

Reviewer 2: ```suggestion
   The nRF9160 DK and Thingy91 come preprovisioned with certificates required to establish a connection to nRF Cloud.
```

>    The default configuration of the :ref:`lib_nrf_cloud` library uses the security tag that the nRF Cloud certificates are stored to.

Reviewer 2: There is something odd here, but I cannot tell what it is. Are the nRF Cloud certificates stored to a security tag? Or are they stored in the device using the security tag? I just don't get the idea here.

Author: You could say that nRF Cloud certificates are stored to a security tag

Configurations for AWS IoT library
----------------------------------

> These options should be set in the overlay configuration file that is used to enable communication with AWS IoT, :file:`overlay-aws.conf`.

Reviewer 2: ```suggestion
To enable communication with AWS IoT, set the following options in the :file:`overlay-aws.conf` file:
```

* :kconfig:`CONFIG_AWS_IOT_BROKER_HOST_NAME`
* :kconfig:`CONFIG_AWS_IOT_SEC_TAG`

Configurations for Azure IoT Hub library
----------------------------------------

> These options should be set in the overlay configuration file that is used to enable communication with Azure IoT Hub, :file:`overlay-azure.conf`.

Reviewer 2: ```suggestion
To enable communication with Azure IoT Hub, set the following options in the :file:`overlay-azure.conf` file:
```

* :kconfig:`CONFIG_AZURE_IOT_HUB_DPS_HOSTNAME`
* :kconfig:`CONFIG_AZURE_IOT_HUB_DPS_ID_SCOPE`
* :kconfig:`CONFIG_AZURE_IOT_HUB_SEC_TAG`
* :kconfig:`CONFIG_AZURE_FOTA_SEC_TAG`

Module hierarchy
****************

The following diagram illustrates the relationship between the cloud module, integration layers, and client libraries.

> .. figure:: /images/asset_tracker_v2_cloud_module_hierarchy.svg
>     :alt: Cloud module hierarchy

Reviewer 2: ```suggestion
.. figure:: /images/asset_tracker_v2_cloud_module_hierarchy.svg
    :alt: Cloud module hierarchy
    
    Cloud module hierarchy
```

Module states
*************

The cloud module has an internal state machine with the following states:

  * ``STATE_LTE_INIT`` - The initial state of the module in which it awaits the modem to be initialized.
  * ``STATE_LTE_DISCONNECTED`` - The module has performed all required initialization and waits for the modem to connect to LTE.
  * ``STATE_LTE_CONNECTED`` - The modem is connected to LTE and the internal cloud connection routine starts. This state has two sub-states:
    * ``SUB_STATE_CLOUD_DISCONNECTED`` - The cloud service is disconnected.
    * ``SUB_STATE_CLOUD_CONNECTED`` - The cloud service is connected, data can now be sent.

>   * ``STATE_SHUTDOWN`` - The module has been shut down after receiving instruction to do so from the util module.

Reviewer 2: ```suggestion
  * ``STATE_SHUTDOWN`` - The module has been shut down after receiving a request to do so from the util module.
```

> State transitions happen based on events from other modules, such as the app module, data module, and util module.

Reviewer 2: ```suggestion
State transitions take place based on events from other modules, such as the app module, data module, and util module.
```

Module events
*************

> The :file:`asset_tracker_v2/src/events/cloud_module_event.h` header file contains a list of various events emitted by the module.

Reviewer 2: Is *emitted* the right verb here? How about initiated, for example?

Author: The header file contains the various events that are sent out by the module. Sent is probably a better word.

Author: Most likely we will have to post a cleanup PR to align all modules that share common sections such as this. So that they use the same wording.

========================

To be able to use a supported cloud client library, you need to set a few mandatory Kconfig options.
These typically include the cloud service hostname and the security tag associated with the certificates used to establish a connection.
Before running the application, you need to provision the certificates to the modem using the same security tag.
For more information on how to set up a connection and provision certificates to the modem, see the documentation for the respective client library in :ref:`Integration layers <integration_layers>`.

.. note::
   There are no mandatory configuration settings for the :ref:`lib_nrf_cloud` library.
   The nRF9160 DK and Thingy91 come preprovisioned with certificates required to establish a connection to nRF Cloud.
   The default configuration of the :ref:`lib_nrf_cloud` library uses the security tag that the nRF Cloud certificates are stored to.

Configurations for AWS IoT library
----------------------------------

To enable communication with AWS IoT, set the following options in the :file:`overlay-aws.conf` file:

* :kconfig:`CONFIG_AWS_IOT_BROKER_HOST_NAME`
* :kconfig:`CONFIG_AWS_IOT_SEC_TAG`

> 

Reviewer 2: Remove extra empty line.

Module hierarchy
****************

The following diagram illustrates the relationship between the cloud module, integration layers, and client libraries.

.. figure:: /images/asset_tracker_v2_cloud_module_hierarchy.svg
    :alt: Cloud module hierarchy

    Cloud module hierarchy

Module states
*************

The cloud module has an internal state machine with the following states:

* ``STATE_LTE_INIT`` - The initial state of the module in which it awaits the modem to be initialized.
* ``STATE_LTE_DISCONNECTED`` - The module has performed all required initialization and waits for the modem to connect to LTE.
* ``STATE_LTE_CONNECTED`` - The modem is connected to LTE and the internal cloud connection routine starts. This state has two sub-states:

   * ``SUB_STATE_CLOUD_DISCONNECTED`` - The cloud service is disconnected.
   * ``SUB_STATE_CLOUD_CONNECTED`` - The cloud service is connected, data can now be sent.

* ``STATE_SHUTDOWN`` - The module has been shut down after receiving a request to do so from the util module.

State transitions take place based on events from other modules, such as the app module, data module, and util module.

Module events
*************

> The :file:`asset_tracker_v2/src/events/cloud_module_event.h` header file contains a list of various events emitted by the module.

Reviewer 2: ```suggestion
The :file:`asset_tracker_v2/src/events/cloud_module_event.h` header file contains a list of various events sent by the module.
```
